fix: upgrade CI to GCC 13 and fix compilation issues, fixes #557

sublimator opened the pull request:
Updates CI workflow to use GCC 13 and fixes compiler issues

## Changes
- Upgraded CI workflow from gcc-11/g++-11 to gcc-13/g++-13
  - Release builder covers gcc 11
- Added missing `#include <cstdint>` header in `src/ripple/app/hook/Enum.h`
- Fixed issue in `SetHook_test.cpp` by explicitly specifying `std::vector<Json::Value>` type
- Removed overly broad cache restore keys that were causing cache pollution between different compiler versions
- Parameterized compiler version in workflow matrix
- Bumped cache version to force clean rebuild

Fixes #557

sublimator committed:
fix: gcc13 compile issues, fixes #557

sublimator commented on src/test/app/SetHook_test.cpp:
```diff
@@ -321,7 +321,7 @@ class SetHook0_test : public beast::unit_test::suite
             ter(temMALFORMED));
 
-        env(ripple::test::jtx::hook(alice, {{}}, 0),
+        env(ripple::test::jtx::hook(alice, std::nullopt, 0),
```
testing 123

RichardAH replied:
as discussed in mm pass empty vec here

sublimator committed:
fix: fix restore key

sublimator committed:
fix: parameterize compiler version in workflow matrix

- Add compiler_version to matrix configuration
- Remove hardcoded version 13 and unnecessary clang conditional
- Use matrix.compiler_version for Conan profile configuration

sublimator committed:
fix: bump cache version to force fresh conan build

Invalidate cached conan packages that have gcc-11 paths baked in

RichardAH committed:
Merge branch 'dev' into niq-xahau-fixes
